python/tchannel/tornado/peer.py

```diff
@@ -458,34 +506,81 @@ def send(self, arg1, arg2, arg3, traceflag=False, headers=None):
             parent_span_id = None
             trace_id = None
 
-        # set 'as' to 'raw if 'as' is empty
+        # set default transport headers
         headers = headers or {}
-        if 'as' not in headers:
-            headers['as'] = 'raw'
-
-        connection = yield self.peer.connect()
-        message_id = connection.next_message_id()
-        response = yield connection.send_request(
-            Request(
-                service=self.service,
-                argstreams=[InMemStream(endpoint), arg2, arg3],
-                id=message_id,
-                headers=headers,
-                tracing=Trace(
-                    name=endpoint,
-                    trace_id=trace_id,
-                    parent_span_id=parent_span_id,
-                    endpoint=Endpoint(self.peer.host,
-                                      self.peer.port,
-                                      self.service),
-                    traceflags=traceflag,
-                )
+        for k, v in self.headers.iteritems():
+            headers.setdefault(k, v)
+
+        original_request = Request(
+            service=self.service,
+            argstreams=[InMemStream(endpoint), arg2, arg3],
+            id=0,
+            headers=headers,
+            tracing=Trace(
+                name=endpoint,
+                trace_id=trace_id,
+                parent_span_id=parent_span_id,
+                endpoint=Endpoint(self.peer.host,
+                                  self.peer.port,
+                                  self.service),
+                traceflags=traceflag,
             )
         )
 
+        response = None
+        # black list to record all used peers, so they aren't chosen again.
+        blacklist = set()
+        # only retry on non-stream request
+        if original_request.is_streaming_request():
+            connection = yield self.peer.connect()
+            message_id = connection.next_message_id()
+            original_request.id = message_id
+            response = yield connection.send_request(original_request)
+        else:
+            req = original_request
+            # mac number of times to retry 5.
+            peer = self.peer
+            for num_of_retry in range(retry_times):
+                try:
+                    req = original_request.clone()
```

Comment:
We don't know if we need to unless we try. So we have to clone first then send, otherwise when we should do retry but we haven't made a copy of request, it will be too late.